File: ts/src/render.ts
/** render.ts — terminal + JSON rendering of a ScanReport. */
import { ScanReport } from "./scan.js";

const TTY = process.stdout.isTTY && !process.env.NO_COLOR;
const c = (code: string, s: string) => (TTY ? `\x1b[${code}m${s}\x1b[0m` : s);
const amber = (s: string) => c("33", s);
const danger = (s: string) => c("91", s);
const safe = (s: string) => c("92", s);
const dim = (s: string) => c("90", s);
const bold = (s: string) => c("1", s);

function bar(pct: number, width = 32): string {
  const fill = Math.max(0, Math.min(width, Math.round((pct / 100) * width)));
  return "█".repeat(fill) + "░".repeat(width - fill);
}

export function renderTerminal(r: ScanReport, top = 5): string {
  const L: string[] = [];
  L.push(bold("mcp-turnstile") + dim(`  ·  ${r.server}  ·  ${r.transport}  ·  ${r.nTools} tools`));
  L.push(dim(`tokenizer: ${r.tokenizer}`));
  L.push("");
  L.push(bold("TOKEN TAX"));
  L.push(`  ${amber(bar(r.pctContext))} ${amber(r.pctContext + "%")} of a ${Math.round(r.window / 1000)}k window`);
  L.push(`  ${r.fullTokens.toLocaleString()} tokens for tool definitions  ·  ${r.ratio}x over a minimal schema  ·  ${r.nTools ? Math.floor(r.fullTokens / r.nTools) : 0} avg/tool`);
  if (r.perTool.length) {
    L.push(dim("  heaviest tools:"));
    for (const t of r.perTool.slice(0, top)) L.push(dim(`    ${String(t.tokens).padStart(6)}  ${t.name}`));
  }
  L.push("");
  if (r.progressive) {
    const p = r.progressive;
    L.push(bold("IF YOU DISCLOSE PROGRESSIVELY"));
    L.push(`  index costs ${p.indexTokens.toLocaleString()} tokens; a task loads one tool for ~${p.taskTokens.toLocaleString()} tokens`);
    if (p.reductionPct > 0) L.push(`  ${safe(p.reductionPct + "% fewer tokens")} per task vs. loading everything`);
    else L.push(dim("  too few tools to save — progressive disclosure would cost more here"));
    L.push("");
  }
  L.push(bold("METADATA SAFETY"));
  if (!r.findings.length) L.push("  " + safe("✓ no concealment, poisoning, or rug-pull signals"));
  else {
    for (const f of r.findings.filter((x) => x.severity === "high")) L.push("  " + danger(`🚩 [${f.kind}] `) + `${f.tool}: ` + dim(f.detail));
    for (const f of r.findings.filter((x) => x.severity !== "high").slice(0, 8)) L.push("  " + dim(`·  [${f.kind}] ${f.tool}: ${f.detail}`));
  }
  L.push("");
  const seccol = r.grades.safety === "A" ? safe : danger;
  L.push(bold("GRADE") + `   efficiency ${amber(r.grades.efficiency)}   ·   safety ${seccol(r.grades.safety)}` + (r.highFindings ? danger(`   (${r.highFindings} high finding${r.highFindings !== 1 ? "s" : ""})`) : ""));
  if (r.rugPulls) L.push(danger(`  ⚠ ${r.rugPulls} tool(s) changed since last scan (rug-pull)`));
  return L.join("\n");
}

export function renderJson(r: ScanReport): string {
  return JSON.stringify(r, null, 2);
}

File: mcpturn/render.py
"""render.py — turn a ScanReport into a colored terminal report or JSON."""

from __future__ import annotations

import json
import os
import sys

from .scan import ScanReport

_TTY = sys.stdout.isatty() and os.environ.get("NO_COLOR") is None


def _c(code: str, s: str) -> str:
    return f"\033[{code}m{s}\033[0m" if _TTY else s


AMBER = lambda s: _c("33", s)
DANGER = lambda s: _c("91", s)
SAFE = lambda s: _c("92", s)
DIM = lambda s: _c("90", s)
BOLD = lambda s: _c("1", s)


def _bar(pct: float, width: int = 32) -> str:
    fill = max(0, min(width, round(pct / 100 * width)))
    return "█" * fill + "░" * (width - fill)


def render_terminal(rep: ScanReport, top: int = 5) -> str:
    L = []
    L.append(BOLD("mcp-turnstile") + DIM(f"  ·  {rep.server}  ·  {rep.transport}  ·  {rep.n_tools} tools"))
    L.append(DIM(f"tokenizer: {rep.tokenizer}"))
    L.append("")

    # token tax
    L.append(BOLD("TOKEN TAX"))
    L.append(f"  {AMBER(_bar(rep.pct_context))} {AMBER(str(rep.pct_context) + '%')} of a {rep.window // 1000}k window")
    L.append(f"  {rep.full_tokens:,} tokens for tool definitions  ·  "
             f"{rep.ratio}x over a minimal schema  ·  "
             f"{rep.full_tokens // rep.n_tools if rep.n_tools else 0} avg/tool")
    if rep.per_tool:
        L.append(DIM("  heaviest tools:"))
        for t in rep.per_tool[:top]:
            L.append(DIM(f"    {t['tokens']:>6,}  {t['name']}"))
    L.append("")

    # progressive potential
    if rep.progressive:
        p = rep.progressive
        L.append(BOLD("IF YOU DISCLOSE PROGRESSIVELY"))
        L.append(f"  index costs {p['index_tokens']:,} tokens; task '{p['example_task']}' loads "
                 f"{AMBER(p['loaded'])} for {p['task_tokens']:,} tokens")
        if p["reduction_pct"] > 0:
            L.append(f"  {SAFE(str(p['reduction_pct']) + '% fewer tokens')} per task vs. loading everything")
        else:
            L.append(DIM("  too few tools to save — progressive disclosure would cost more here"))
        L.append("")

    # safety
    L.append(BOLD("METADATA SAFETY"))
    if not rep.findings:
        L.append("  " + SAFE("✓ no concealment, poisoning, or rug-pull signals"))
    else:
        highs = [f for f in rep.findings if f["severity"] == "high"]
        lows = [f for f in rep.findings if f["severity"] != "high"]
        for f in highs:
            L.append("  " + DANGER(f"🚩 [{f['kind']}] ") + f"{f['tool']}: " + DIM(f["detail"]))
        for f in lows[:8]:
            L.append("  " + DIM(f"·  [{f['kind']}] {f['tool']}: {f['detail']}"))
        if len(lows) > 8:
            L.append(DIM(f"  … +{len(lows) - 8} more informational"))
    L.append("")

    # grades
    g = rep.grades
    seccol = SAFE if g.get("safety") == "A" else DANGER
    L.append(BOLD("GRADE") + f"   efficiency {AMBER(g.get('efficiency','?'))}   ·   safety {seccol(g.get('safety','?'))}"
             + (DANGER(f"   ({rep.high} high finding{'s' if rep.high != 1 else ''})") if rep.high else ""))
    if rep.rug_pulls:
        L.append(DANGER(f"  ⚠ {rep.rug_pulls} tool(s) changed since last scan (rug-pull)"))
    return "\n".join(L)


def render_json(rep: ScanReport) -> str:
    return json.dumps(rep.to_dict(), indent=2, ensure_ascii=False)
